File: DeepspaceRuby/lib/Dice.rb
#encoding: UTF-8

require_relative 'GameCharacter'

module Deepspace
    class Dice
        @NHANGARSPROB
        @NSHIELDSPROB
        @NWEAPONSPROB
        @FIRSTSHOTPROB
        @EXTRAEFFICIENCYPROB
        @generator
        
        def initialize
            @NHANGARSPROB = 0.25
            @NSHIELDSPROB = 0.25
            @NWEAPONSPROB = 0.33
            @FIRSTSHOTPROB = 0.5
            @EXTRAEFFICIENCYPROB=0.8
            @generator = Random.new
        end
        
        def initWithNHangars
            r = @generator.rand(1.0)
            if r < @NHANGARSPROB
                return 0
            else
                return 1
            end
        end

        def initWithNWeapons
            r = @generator.rand(1.0)
            if r < @NWEAPONSPROB
                return 0
            elsif ((r > @NWEAPONSPROB) && (r < 2*@NWEAPONSPROB))
                return 1
            else
                return 2
            end
        end

        def initWithNShields
            r = @generator.rand(1.0)
            if r < @NSHIELDSPROB
                return 0
            else
                return 1
            end
        end

        def whoStarts nPlayers
            return @generator.rand(nPlayers)
        end

        def firstShot
            r = @generator.rand(1.0)
            if r < @FIRSTSHOTPROB
                return GameCharacter::SPACESTATION
            else
                return GameCharacter::ENEMYSTARSHIP
            end
        end

        def spaceStationMoves(speed)
            r = @generator.rand(1.0)
            return r < speed
        end

        def extraEfficiency
            r = @generator.rand(1.0)
            return r < @EXTRAEFFICIENCYPROB
        end

        def to_s
      
            initnHangars = "- INIT WITH NHANGARS: #{initWithNHangars} "
            initnWeapons = "- INIT WITH NWEAPONS: #{initWithNWeapons} "
            initnShields = "- INIT WITH NSHIELDS: #{initWithNShields}\n"
            if firstShot == Deepspace::GameCharacter::SPACESTATION
              firstS = "- FIRST SHOT: SPACESTATION\n"
            else
              firstS = "- FIRST SHOT: ENEMYSTARSHIP\n"
            end
            
            return initnHangars + initnWeapons + initnShields + firstS
        end
    end
end

File: deepspace/Dice.java
/*
 * @author: Jesús García León
 * @file: Dice
 */
package deepspace;

import java.util.Random;

/**
 * @brief Tiene la responsabilidad de tomar todas las 
 *        decisiones que dependen del azar en el juego
 */
public class Dice {
    private final float NHANGARSPROB;
    private final float NSHIELDSPROB;
    private final float NWEAPONSPROB;
    private final float FIRSTSHOTPROB;
    private final float EXTRAEFFICIENCYPROB;
    Random generator;
    
    Dice(){
        NHANGARSPROB = 0.25f;
        NSHIELDSPROB = 0.25f;
        NWEAPONSPROB = 0.33f;
        FIRSTSHOTPROB = 0.5f;
        EXTRAEFFICIENCYPROB = 0.8f;
        generator = new Random();
    }
    
    public int initWithNHangars(){
        float r = generator.nextFloat();
        if(r < NHANGARSPROB)
            return 0;
        else
            return 1;
    }
    
    public int initWithNWeapons(){
        float r = generator.nextFloat();
        if(r < NWEAPONSPROB)
            return 1;
        else if(r >= NWEAPONSPROB && r < 2*NWEAPONSPROB)
            return 2;
        else 
            return 3;
    }
    
    public int initWithNShiels(){
        float r = generator.nextFloat();
        if(r < NSHIELDSPROB)
            return 0;
        else
            return 1;
    }
    
    public int whoStarts(int nPlayers){
        return generator.nextInt(nPlayers);
    }

    public GameCharacter firstShot(){
        float r = generator.nextFloat();
        if(r <= FIRSTSHOTPROB)
            return GameCharacter.SPACESTATION;
        else
            return GameCharacter.ENEMYSTARSHIP;
    }
    
    public boolean spaceStationMoves(float speed){
        float r = generator.nextFloat();
        return r <= speed;
    }
    
    public boolean extraEfficiency(){
        float r = generator.nextFloat();
        return  r < EXTRAEFFICIENCYPROB;
    }
}
